scio-smb/src/main/java/org/apache/beam/sdk/extensions/smb/AvroBucketMetadata.java

```diff
@@ -108,6 +139,100 @@ public boolean isPartitionCompatible(BucketMetadata o) {
         && Arrays.equals(keyPath, that.keyPath);
   }
 
+  private static String validateKeyField(String keyField, Class keyClass, Schema schema) {
+    final String[] keyPath = toKeyPath(keyField);
+
+    Schema currSchema = schema;
+    for (int i = 0; i < keyPath.length - 1; i++) {
+      final Schema.Field field = currSchema.getField(keyPath[i]);
+      Preconditions.checkNotNull(
+          field, String.format("Key path %s does not exist in schema %s", keyPath[i], currSchema));
+
+      currSchema = field.schema();
+      Preconditions.checkArgument(
+          currSchema.getType() == Schema.Type.RECORD,
+          "Non-leaf key field " + keyPath[i] + " is not a Record type");
+    }
+
+    final Schema.Field finalKeyField = currSchema.getField(keyPath[keyPath.length - 1]);
+    Preconditions.checkNotNull(
+        finalKeyField,
+        String.format(
+            "Leaf key field %s does not exist in schema %s",
+            keyPath[keyPath.length - 1], currSchema));
+
+    final Class<?> finalKeyFieldClass = getKeyClassFromSchema(finalKeyField.schema());
+
+    Preconditions.checkArgument(
+        finalKeyFieldClass.isAssignableFrom(keyClass),
+        String.format(
+            "Key class %s did not conform to its Avro schema. Must be of class: %s",
+            keyClass, finalKeyFieldClass));
+
+    return keyField;
+  }
+
+  private static Class<?> getKeyClassFromSchema(Schema schema) {
+    Schema.Type schemaType = schema.getType();
+
+    if (schemaType == Schema.Type.UNION) {
+      boolean hasNull = false;
+
+      for (Schema typeSchema : schema.getTypes()) {
+        if (typeSchema.getType() == Schema.Type.NULL) {
+          hasNull = true;
+        } else {
+          schemaType = typeSchema.getType();
+        }
+      }
+
+      Preconditions.checkArgument(
+          hasNull && schema.getTypes().size() == 2,
+          "A union can only be used as a key field if it contains exactly 1 null "
+              + "type and 1 key-class type. Was: "
+              + schema);
+    }
+
+    return getClassForType(schemaType);
+  }
+
+  private static Class<?> getClassForType(Schema.Type schemaType) {
```

Comment:
Surprised that Avro doesn't already have a mapping table like this? Or maybe it's private?

Reply:
Sort of... it has `SpecificData#getClass`:

https://github.com/apache/avro/blob/release-1.8.2/lang/java/avro/src/main/java/org/apache/avro/specific/SpecificData.java#L152

But it doesn't support Fixed or Record; behaves strangely for union types (`"type": "string"` returns `String.class` ; whereas `"type":["string", "null"]` returns `CharSequence.class`); and returns null for Enum types unless there's a Java class of the same name in the classpath*. I guess we could check for `Record`, `Fixed`, `Enum`, and `String` types ourselves first and then delegate to `SpecificData#getClass`.

* Which actually, we could maybe make a ticket to support--since right now `extractKey` just works by casting the raw value from the generic record to type `K`: `K key = (K) node.get(keyPath[keyPath.length - 1]);`. We could have some logic that if the key schema is an enum we could use reflection on the enum class like `EnumClass.ValueOf((String) node.get(keyPath[keyPath.length - 1])`. but yeah... not sure if it's a big enough use case to matter.

Reply:
If enum is the only edge case right now probably not worth it. Enum doesn't have enough range and shouldn't be used as key anyway.